[SPARK-18189] [SQL] Fix serialization issue in KeyValueGroupedDataset

seyfe opened the pull request:
## What changes were proposed in this pull request?
Likewise [DataSet.scala](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L156) KeyValueGroupedDataset should mark the queryExecution as transient. 

As mentioned in the Jira ticket, without @transient we saw serialization issues like 

```
Caused by: java.io.NotSerializableException: org.apache.spark.sql.execution.QueryExecution
Serialization stack:
        - object not serializable (class: org.apache.spark.sql.execution.QueryExecution, value: ==
```

## How was this patch tested?

Run the query which is specified in the Jira ticket before and after:
```
val a = spark.createDataFrame(sc.parallelize(Seq((1,2),(3,4)))).as[(Int,Int)]
val grouped = a.groupByKey(
{x:(Int,Int)=>x._1}
)
val mappedGroups = grouped.mapGroups((k,x)=>
{(k,1)}
)
val yyy = sc.broadcast(1)
val last = mappedGroups.rdd.map(xx=>
{ val simpley = yyy.value 1 }
)
```

seyfe committed:
Make queryExecution transient for KeyValueGroupedDataset

AmplabJenkins commented:
Can one of the admins verify this patch?

hvanhovell commented:
Ok to test

rxin commented:
This guy @transient must get a lot of emails ...

rxin commented:
Can you change the title to "[SQL]" rather than "[core]"?

rxin commented:
LGTM pending Jenkins.

SparkQA commented:
**[Test build #3383 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3383/consoleFull)** for PR 15706 at commit [`07160b9`](https://github.com/apache/spark/commit/07160b93130ba58c29e516368354fea86e5e8865).

SparkQA commented:
**[Test build #3383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3383/consoleFull)** for PR 15706 at commit [`07160b9`](https://github.com/apache/spark/commit/07160b93130ba58c29e516368354fea86e5e8865).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented:
Actually is there a test case you can create for this? Make sure the test case would fail when the patch is not applied.

SparkQA commented:
**[Test build #3385 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3385/consoleFull)** for PR 15706 at commit [`07160b9`](https://github.com/apache/spark/commit/07160b93130ba58c29e516368354fea86e5e8865).

SparkQA commented:
**[Test build #3385 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3385/consoleFull)** for PR 15706 at commit [`07160b9`](https://github.com/apache/spark/commit/07160b93130ba58c29e516368354fea86e5e8865).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

seyfe committed:
Add unittest in ReplSuite Scala-2.11

seyfe commented:
Hi @rxin ,

I added a unittest in [DatasetSuite](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala) but I wasn't able to repro it. Bug only reproduce in spark-shell. That's why I moved the unittest to [ReplSuite](https://github.com/apache/spark/blob/master/repl/scala-2.11/src/test/scala/org/apache/spark/repl/ReplSuite.scala).

I checked the history and seems like we are adding unittests to only scala-2.11 now. If that's not the case, I can create same one in [ReplSuite scala-210](https://github.com/apache/spark/blob/master/repl/scala-2.10/src/test/scala/org/apache/spark/repl/ReplSuite.scala)

Please let me know.

Thanks

SparkQA commented:
**[Test build #3388 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3388/consoleFull)** for PR 15706 at commit [`a9b089a`](https://github.com/apache/spark/commit/a9b089ac9e45e9454bc0b6fa592cc0d624be99a2).

rxin commented:
Thanks! LGTM pending Jenkins.

SparkQA commented:
**[Test build #3388 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3388/consoleFull)** for PR 15706 at commit [`a9b089a`](https://github.com/apache/spark/commit/a9b089ac9e45e9454bc0b6fa592cc0d624be99a2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented:
Merging in master/branch-2.0.
 Thanks!

srowen commented:
I think this might be causing the master/2.0 builds (but only Maven, weirdly) to fail. This includes the 2.0.2 RC2, which I built with Maven. See for example:

https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-2.2/2200/
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-2.0-test-maven-hadoop-2.7/855/

rxin commented:
@srowen that's weird - do you have time to dig into this a bit?

srowen commented:
Yeah will dig in, may not have an answer until tomorrow. If you're looking to push ahead 2.0.2 I'd just revert in branch-2.0 for now.

srowen commented:
@seyfe so my guess is that, not surprisingly, this is some pretty arcane interaction between several of a) codegen, b) Netty's class hierarchy, c) SBT vs Maven classloading and d) the spark-shell's classloading peculiarities.

I strongly suspect it's an issue with the test of course, not the underlying change. I wonder if we can just write a different test that doesn't trip this.

Is it necessary to test this in the REPL? can we just write the code it tests as a test directly?

seyfe commented:
Hi @srowen. I initially put the test in DatasetSuite but the test was passing with and without my change. Since, I was able to repro only in spark-shell I moved the test to REPL suite.

I didn't run the test in branch-2.0 only run in master. Let me check if we can fix it so, it would work on branch-2.0 as well.

seyfe commented:
I run the test in branch-2.0 locally and it works fine for me.
Seems like related with https://github.com/apache/spark/pull/15286. I need to dig in more.

srowen commented:
I think it may only involve the same exception but I am not sure it is the same. 

I don't like removing tests of course but I think I conclude that the issue you intend to fix is fixed. It is the test that has a problem. We do want to make the change but need a test that will pass. If there is no obvious workaround for what looks to me like a fairly obscure class loader issue, then we could make this test not use the REPL.

It would still test something useful and catch some potential regressions. It would not evidently catch this one though I suppose it is unlikely this exact issue would manifest again.  

I could live with a simpler test to make this work and get the change into 2.0.2 because otherwise it has to be reverted for now.

seyfe commented:
@srowen, that is fair. Let me move this test from Repl into DatasetSuite for branch-2.0 and send a PR.

seyfe commented:
Hi @srowen ,

Here is the PR:
https://github.com/apache/spark/pull/15774
